# Module 1 : Pre-processing, Analyzing data using Python and SQL

## Task 1 : Pre-processing the data

In [1]:
# Firstly, we would have to import all the necessary libraries to be utilized

import pandas as pd
import numpy as np
from numpy import nan
from datetime import datetime, timedelta

In [2]:
#There are two datasets which we would then import using pandas
apps = pd.read_csv("playstore_apps.csv", index_col = 'App')
reviews = pd.read_csv("playstore_reviews.csv")

In [3]:
#Lets see the column attributes for apps dataset
apps.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10841 entries, Photo Editor & Candy Camera & Grid & ScrapBook to iHoroscope - 2018 Daily Horoscope & Astrology
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Category        10841 non-null  object 
 1   Rating          9367 non-null   float64
 2   Reviews         10840 non-null  float64
 3   Size            10841 non-null  object 
 4   Installs        10840 non-null  float64
 5   Type            10840 non-null  object 
 6   Price           10840 non-null  float64
 7   Content Rating  10840 non-null  object 
 8   Genres          10841 non-null  object 
 9   Last Updated    10840 non-null  object 
 10  Current Ver     10833 non-null  object 
 11  Android Ver     10838 non-null  object 
dtypes: float64(4), object(8)
memory usage: 1.1+ MB


In [4]:
# Let's find the no of rows and columns in apps dataset
apps.shape

(10841, 12)

In [5]:
# Let's view the first 5 rows of the apps dataset
apps.head()

Category  Rating  \
App                                                                          
Photo Editor & Candy Camera & Grid & ScrapBook      ART_AND_DESIGN     4.1   
Coloring book moana                                 ART_AND_DESIGN     3.9   
U Launcher Lite – FREE Live Cool Themes, Hide Apps  ART_AND_DESIGN     4.7   
Sketch - Draw & Paint                               ART_AND_DESIGN     4.5   
Pixel Draw - Number Art Coloring Book               ART_AND_DESIGN     4.3   

                                                     Reviews  Size  \
App                                                                  
Photo Editor & Candy Camera & Grid & ScrapBook         159.0   19M   
Coloring book moana                                    967.0   14M   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   87510.0  8.7M   
Sketch - Draw & Paint                               215644.0   25M   
Pixel Draw - Number Art Coloring Book                  967.0  2.8M   

                                                      Installs  Type  Price  \
App                                                                           
Photo Editor & Candy Camera & Grid & ScrapBook         10000.0  Free    0.0   
Coloring book moana                                   500000.0  Free    0.0   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   5000000.0  Free    0.0   
Sketch - Draw & Paint                               50000000.0  Free    0.0   
Pixel Draw - Number Art Coloring Book                 100000.0  Free    0.0   

                                                   Content Rating  \
App                                                                 
Photo Editor & Candy Camera & Grid & ScrapBook           Everyone   
Coloring book moana                                      Everyone   
U Launcher Lite – FREE Live Cool Themes, Hide Apps       Everyone   
Sketch - Draw & Paint                                        Teen   
Pixel Draw - Number Art Coloring Book                    Everyone   

                                                                       Genres  \
App                                                                             
Photo Editor & Candy Camera & Grid & ScrapBook                   Art & Design   
Coloring book moana                                 Art & Design;Pretend Play   
U Launcher Lite – FREE Live Cool Themes, Hide Apps               Art & Design   
Sketch - Draw & Paint                                            Art & Design   
Pixel Draw - Number Art Coloring Book                 Art & Design;Creativity   

                                                   Last Updated  \
App                                                               
Photo Editor & Candy Camera & Grid & ScrapBook       07-01-2018   
Coloring book moana                                  15-01-2018   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   01-08-2018   
Sketch - Draw & Paint                                08-06-2018   
Pixel Draw - Number Art Coloring Book                20-06-2018   

                                                           Current Ver  \
App                                                                      
Photo Editor & Candy Camera & Grid & ScrapBook                   1.0.0   
Coloring book moana                                              2.0.0   
U Launcher Lite – FREE Live Cool Themes, Hide Apps               1.2.4   
Sketch - Draw & Paint                               Varies with device   
Pixel Draw - Number Art Coloring Book                              1.1   

                                                     Android Ver  
App                                                               
Photo Editor & Candy Camera & Grid & ScrapBook      4.0.3 and up  
Coloring book moana                                 4.0.3 and up  
U Launcher Lite – FREE Live Cool Themes, Hide Apps  4.0.3 and up  
Sketch - Draw & Paint                                 4.2 and up  
Pixel Draw - Number Art

In [6]:
apps.duplicated().value_counts()

False    10349
True       492
dtype: int64

### Subtask 1 : Removing Duplicate Rows

We have established that 483 columns in the apps dataset contains duplicate rows, while the App column contains 9660 unique columns. we will then attempt to remove those duplicate rows from our apps dataset.

In [7]:
# Drop duplicate rows in 'App' column
apps = apps.drop_duplicates()

In [8]:
#Check to make sure duplicates rows were dropped
apps.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10349 entries, Photo Editor & Candy Camera & Grid & ScrapBook to iHoroscope - 2018 Daily Horoscope & Astrology
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Category        10349 non-null  object 
 1   Rating          8891 non-null   float64
 2   Reviews         10348 non-null  float64
 3   Size            10349 non-null  object 
 4   Installs        10348 non-null  float64
 5   Type            10348 non-null  object 
 6   Price           10348 non-null  float64
 7   Content Rating  10348 non-null  object 
 8   Genres          10349 non-null  object 
 9   Last Updated    10348 non-null  object 
 10  Current Ver     10341 non-null  object 
 11  Android Ver     10346 non-null  object 
dtypes: float64(4), object(8)
memory usage: 1.0+ MB


### SubTask 2: Remove Irrelevant values from each column if any

We would check each column to ascertain attributes that are irrelevant and must be removed from the dataset.

We would start by checking the unique attributes for each column so as to source for irregularities.

In [9]:
apps['Category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

The `category` column loos okay except the 1.9 attribute as seen above which is not a google playstore category and thus we would attempt to remove the row entry containing the category of '1.9'

In [10]:
#Check the row that contains the category of 1.9
apps[apps['Category'] == '1.9']

Category  Rating  Reviews    Size  \
App                                                                         
Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0      NaN  1,000+   

                                         Installs Type  Price Content Rating  \
App                                                                            
Life Made WI-Fi Touchscreen Photo Frame       NaN    0    NaN            NaN   

                                                    Genres Last Updated  \
App                                                                       
Life Made WI-Fi Touchscreen Photo Frame  February 11, 2018          NaN   

                                        Current Ver Android Ver  
App                                                              
Life Made WI-Fi Touchscreen Photo Frame  4.0 and up         NaN

In [11]:
# only one row (row 10472) contain the category of 1.9 and that row should be removed form our dataset since its not needed.

apps.drop('Life Made WI-Fi Touchscreen Photo Frame', inplace=True)

In [12]:
# Check to make sure the category of 1.9 was removed.

apps['Category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION'],
      dtype=object)

In [13]:
# Check the distinct attributes of `Rating` column
apps['Rating'].unique()

array([4.1, 3.9, 4.7, 4.5, 4.3, 4.4, 3.8, 4.2, 4.6, 3.2, 4. , nan, 4.8,
       4.9, 3.6, 3.7, 3.3, 3.4, 3.5, 3.1, 5. , 2.6, 3. , 1.9, 2.5, 2.8,
       2.7, 1. , 2.9, 2.3, 2.2, 1.7, 2. , 1.8, 2.4, 1.6, 2.1, 1.4, 1.5,
       1.2])

The `Rating` column contains missing values. Since the Ratings for each app is a numerical value, these missing values will be replaced with '0' much later in our analysis using Microsoft Excel after our dataset has been exported to csv.

In [14]:
# Check the distinct attributes of `Content Rating` column
apps['Content Rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [15]:
# Check for unique column attributes
apps['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

#### There is an unwanted `nan` row on the `Type` column which should be removed as it's not needed

In [16]:
# Let's remove the nan rows in the `Type` column
apps = apps.dropna(subset=['Type'])

#### Let's confirm that the nan row has been removed from the `Type` column

In [17]:
apps['Type'].unique()

array(['Free', 'Paid'], dtype=object)

### Next is to check for incorrect datatype in column attributes. The `Last Updated` column has a datatype of object and must be changed to a datatype of datetime to reflect the date attributes in it

In [18]:
# Change datatype of `Last Updated` column to datetime

apps['Last Updated'] = pd.to_datetime(apps['Last Updated'])

C:\Users\kennedy\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '15-01-2018' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\kennedy\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '20-06-2018' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\kennedy\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '26-03-2017' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\kennedy\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '26-04-2018' in DD/MM/YYYY format. Provide format o

In [19]:
# Let's check that all changes made are reflected in every column.
apps.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10347 entries, Photo Editor & Candy Camera & Grid & ScrapBook to iHoroscope - 2018 Daily Horoscope & Astrology
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Category        10347 non-null  object        
 1   Rating          8890 non-null   float64       
 2   Reviews         10347 non-null  float64       
 3   Size            10347 non-null  object        
 4   Installs        10347 non-null  float64       
 5   Type            10347 non-null  object        
 6   Price           10347 non-null  float64       
 7   Content Rating  10347 non-null  object        
 8   Genres          10347 non-null  object        
 9   Last Updated    10347 non-null  datetime64[ns]
 10  Current Ver     10339 non-null  object        
 11  Android Ver     10345 non-null  object        
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 1.0+ MB


From the above assessment, it can be seen that the `app` dataset still contains missing values in the `Rating` , `Current Ver` and `Andriod Ver` columns which will be addressed in the next stage of our analysis using Microsoft Excel.

### Now we can export our `apps` dataset to csv file for further accessment on microsoft Excel

In [20]:
apps.to_csv('cleaned_apps_v2.csv')

## Pre-processing on the `review` dataset

Lets programmatically view the reviews dataset to get familiar with its properties.

In [21]:
# Lets see the column attributes for reviews dataset
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [22]:
reviews.head(10)

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   
5  10 Best Foods for You                                           Best way   
6  10 Best Foods for You                                            Amazing   
7  10 Best Foods for You                                                NaN   
8  10 Best Foods for You                               Looking forward app,   
9  10 Best Foods for You              It helpful site ! It help foods get !   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000  
5  Positive                1.00                0.300000  
6  Positive                0.60                0.900000  
7       NaN                 NaN                     NaN  
8   Neutral                0.00                0.000000  
9   Neutral                0.00                0.000000

In [23]:
reviews['Translated_Review'].isna().value_counts()

False    37427
True     26868
Name: Translated_Review, dtype: int64

We can observed that the `reviews` dataset has 26868 null values in its `Translated_Review` column. Therefore all rows with null values in the `Translated_Review` column will be deleted as there is no translated reviews present and as such that kind of data is of no use for our analysis
    
The reviews dataset will be further cleaned on Microsoft Excel for both null values and duplicate values before exporting for analysis.

    All pre-processing on the reviews dataset would be carried out in Microsoft Excel.